File: CS/ImageMemoEdit/CustomGridControl/CustomGridView/CustomGridView.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraEditors.Repository;
using EditorDescendant;
using DevExpress.XtraGrid.Views.Printing;

namespace ImageTextEdit
{
    [System.ComponentModel.DesignerCategory("")]
    public class CustomGridView : GridView
    {
        GridCellInfo nowDrawingCellInfo;
        public CustomGridView() : base() { }
        public CustomGridView(GridControl grid) : base(grid) { }

        private static readonly object _onSmartIconSelection = new object();

        public override void Assign(BaseView v, bool copyEvents)
        {
            if (v == null) return;
            BeginUpdate();
            try
            {
                CustomGridView gv = v as CustomGridView;
                if (gv != null)
                    if (copyEvents)
                        Events.AddHandler(_onSmartIconSelection, gv.Events[_onSmartIconSelection]);
            }
            finally { EndUpdate(); }
        }
        public event OnSmartIconSelectionEventHandler OnSmartIconSelection
        {
            add { this.Events.AddHandler(_onSmartIconSelection, value); }
            remove { this.Events.RemoveHandler(_onSmartIconSelection, value); }
        }
        protected internal virtual void RaiseOnSmartIconSelection(SmartIconSelectionEventArgs e)
        {
            OnSmartIconSelectionEventHandler handler = (OnSmartIconSelectionEventHandler)Events[_onSmartIconSelection];
            if (handler != null) handler(this, e);
        }
        protected override string ViewName { get { return "CustomGridView"; } }
        protected override void ActivateEditor(GridCellInfo cell)
        {
            IIconSelector iconSelector = cell.Editor as IIconSelector;
            if (iconSelector != null)
            {
                iconSelector.OnIconSelection += new EditorDescendant.MyRepositoryItemMemoEdit.IconSelectionEventHandler(iconSelector_OnIconSelection);
                nowDrawingCellInfo = cell;
            }
            base.ActivateEditor(cell);
        }

        public override void HideEditor()
        {
            if (!IsEditing || !fAllowCloseEditor) return;
            if (ActiveEditor != null && GridControl != null)
            {
                IIconSelector iconSelector = ActiveEditor.Properties as IIconSelector;
                if (iconSelector != null)
                    iconSelector.OnIconSelection -= new MyRepositoryItemMemoEdit.IconSelectionEventHandler(iconSelector_OnIconSelection);
            }
            base.HideEditor();
        }

        protected override BaseViewPrintInfo CreatePrintInfoInstance(PrintInfoArgs args)
        {
            BaseViewPrintInfo info = new MyGridViewPrintInfo(args);
            return info;
        }

        internal void iconSelector_OnIconSelection(object sender, IconSelectionEventArgs e)
        {
            SmartIconSelectionEventArgs ee = new SmartIconSelectionEventArgs(e.Image, nowDrawingCellInfo, true);
            RaiseOnSmartIconSelection(ee);
            e.Image = ee.Image;
        }
    }
}

File: VB/ImageMemoEdit/CustomGridControl/CustomGridView/CustomGridView.vb
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Views.Base
Imports EditorDescendant
Imports DevExpress.XtraGrid.Views.Printing

Namespace ImageTextEdit

    <System.ComponentModel.DesignerCategory("")>
    Public Class CustomGridView
        Inherits GridView

        Private nowDrawingCellInfo As GridCellInfo

        Public Sub New()
            MyBase.New()
        End Sub

        Public Sub New(ByVal grid As GridControl)
            MyBase.New(grid)
        End Sub

        Private Shared ReadOnly _onSmartIconSelection As Object = New Object()

        Public Overrides Sub Assign(ByVal v As BaseView, ByVal copyEvents As Boolean)
            If v Is Nothing Then Return
            BeginUpdate()
            Try
                Dim gv As CustomGridView = TryCast(v, CustomGridView)
                If gv IsNot Nothing Then
                    If copyEvents Then Events.AddHandler(_onSmartIconSelection, gv.Events(_onSmartIconSelection))
                End If
            Finally
                EndUpdate()
            End Try
        End Sub

        Public Custom Event OnSmartIconSelection As OnSmartIconSelectionEventHandler
            AddHandler(ByVal value As OnSmartIconSelectionEventHandler)
                Events.AddHandler(_onSmartIconSelection, value)
            End AddHandler

            RemoveHandler(ByVal value As OnSmartIconSelectionEventHandler)
                Events.RemoveHandler(_onSmartIconSelection, value)
            End RemoveHandler

            RaiseEvent(ByVal sender As Object, ByVal e As SmartIconSelectionEventArgs)
            End RaiseEvent
        End Event

        Protected Friend Overridable Sub RaiseOnSmartIconSelection(ByVal e As SmartIconSelectionEventArgs)
            Dim handler As OnSmartIconSelectionEventHandler = CType(Events(_onSmartIconSelection), OnSmartIconSelectionEventHandler)
            If handler IsNot Nothing Then handler(Me, e)
        End Sub

        Protected Overrides ReadOnly Property ViewName As String
            Get
                Return "CustomGridView"
            End Get
        End Property

        Protected Overrides Sub ActivateEditor(ByVal cell As GridCellInfo)
            Dim iconSelector As IIconSelector = TryCast(cell.Editor, IIconSelector)
            If iconSelector IsNot Nothing Then
                AddHandler iconSelector.OnIconSelection, New MyRepositoryItemMemoEdit.IconSelectionEventHandler(AddressOf iconSelector_OnIconSelection)
                nowDrawingCellInfo = cell
            End If

            MyBase.ActivateEditor(cell)
        End Sub

        Public Overrides Sub HideEditor()
            If Not IsEditing OrElse Not fAllowCloseEditor Then Return
            If ActiveEditor IsNot Nothing AndAlso GridControl IsNot Nothing Then
                Dim iconSelector As IIconSelector = TryCast(ActiveEditor.Properties, IIconSelector)
                If iconSelector IsNot Nothing Then RemoveHandler iconSelector.OnIconSelection, New MyRepositoryItemMemoEdit.IconSelectionEventHandler(AddressOf iconSelector_OnIconSelection)
            End If

            MyBase.HideEditor()
        End Sub

        Protected Overrides Function CreatePrintInfoInstance(ByVal args As PrintInfoArgs) As BaseViewPrintInfo
            Dim info As BaseViewPrintInfo = New MyGridViewPrintInfo(args)
            Return info
        End Function

        Friend Sub iconSelector_OnIconSelection(ByVal sender As Object, ByVal e As IconSelectionEventArgs)
            Dim ee As SmartIconSelectionEventArgs = New SmartIconSelectionEventArgs(e.Image, nowDrawingCellInfo, True)
            RaiseOnSmartIconSelection(ee)
            e.Image = ee.Image
        End Sub
    End Class
End Namespace
